// ghastly/lib/resolvers/CommandObjectResolver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _BaseResolver = require('./BaseResolver');

var _BaseResolver2 = _interopRequireDefault(_BaseResolver);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @desc Options resolver for `CommandObject` class.
 * @extends BaseResolver
 * @ignore
 */
class CommandObjectResolver extends _BaseResolver2.default {
  /**
   * Constructor.
   */
  constructor() {
    super();

    this.resolver.setRequired(['handler', 'triggers']).setDefaults({
      parameters: [],
      group: null,
      description: null,
      dependencies: [],
      middleware: []
    }).setAllowedTypes('handler', 'function').setAllowedTypes('triggers', 'array').setAllowedTypes('parameters', 'array').setAllowedTypes('group', ['string', 'null']).setAllowedTypes('description', ['string', 'null']).setAllowedTypes('dependencies', ['array', 'object']).setAllowedTypes('middleware', 'array');
  }
}
exports.default = CommandObjectResolver;
