// ghastly/lib/resolvers/ParameterResolver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _BaseResolver = require('./BaseResolver');

var _BaseResolver2 = _interopRequireDefault(_BaseResolver);

var _Types = require('../command/parsers/Types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @desc Options resolver for command parameter definitions.
 * @extends BaseResolver
 * @ignore
 */
class ParameterResolver extends _BaseResolver2.default {
  /**
   * Constructor.
   */
  constructor() {
    super();

    this.resolver.setRequired('name').setDefaults({
      description: null,
      optional: false,
      type: _Types.STRING,
      repeatable: false,
      literal: false,
      defaultValue: options => options.repeatable ? [] : null
    }).setAllowedTypes('name', 'string').setAllowedTypes('description', ['string', 'null']).setAllowedTypes('optional', 'boolean').setAllowedTypes('type', 'string').setAllowedTypes('repeatable', 'boolean').setAllowedTypes('literal', 'boolean');
  }
}
exports.default = ParameterResolver;
