// ghastly/lib/command/parsers/Types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.convertType = exports.isType = exports.resolveType = exports.STR = exports.STRING = exports.NUM = exports.NUMBER = exports.INT = exports.INTEGER = exports.BOOL = exports.BOOLEAN = undefined;

var _lang = require('lodash/lang');

/**
 * Boolean type.
 * @type {string}
 * @const
 * @ignore
 */
const BOOLEAN = exports.BOOLEAN = 'boolean';

/**
 * Alias for Boolean type.
 * @type {string}
 * @const
 * @ignore
 */
const BOOL = exports.BOOL = 'bool';

/**
 * Integer type.
 * @type {string}
 * @const
 * @ignore
 */
const INTEGER = exports.INTEGER = 'integer';

/**
 * Alias for Integer type.
 * @type {string}
 * @const
 * @ignore
 */
const INT = exports.INT = 'int';

/**
 * Number type.
 * @type {string}
 * @const
 * @ignore
 */
const NUMBER = exports.NUMBER = 'number';

/**
 * Alias for Number type.
 * @type {string}
 * @const
 * @ignore
 */
const NUM = exports.NUM = 'num';

/**
 * String type.
 * @type {string}
 * @const
 * @ignore
 */
const STRING = exports.STRING = 'string';

/**
 * Alias for String type.
 * @type {string}
 * @const
 * @ignore
 */
const STR = exports.STR = 'str';

/**
 * Resolves a string value to a type string.
 * @param {string} value - The string.
 * @returns {?string} The resolved value type, or `null` if it could not be resolved.
 * @ignore
 */
const resolveType = exports.resolveType = value => {
  if (!(0, _lang.isString)(value)) {
    throw new TypeError('Expected value to be a string.');
  }

  switch (value.toLowerCase()) {
    case BOOLEAN:
    case BOOL:
      return 'boolean';
    case INTEGER:
    case INT:
      return 'integer';
    case NUMBER:
    case NUM:
      return 'number';
    case STRING:
    case STR:
      return 'string';
    default:
      return null;
  }
};

/**
 * Determines if a string value is of the expected type.
 * @param {string} value - The string.
 * @param {string} expectedType - The expected type.
 * @returns {boolean} `true` if the value is of the expected type, else `false`.
 * @ignore
 */
const isType = exports.isType = (value, expectedType) => {
  if (!(0, _lang.isString)(value)) {
    throw new TypeError('Expected value to be a string.');
  }

  switch (resolveType(expectedType)) {
    case BOOLEAN:
      {
        const lower = value.toLowerCase();

        return lower === 'true' || lower === 'false';
      }
    case INTEGER:
      return !isNaN(value) && (0, _lang.isNumber)(+value);
    case NUMBER:
      return !isNaN(value) && (0, _lang.isNumber)(+value);
    case STRING:
      return true;
    default:
      throw new Error(`Unrecognized type: ${expectedType}.`);
  }
};

/**
 * Converts a string value to a given type.
 * @param {string} value - The string.
 * @param {string} type - The type to convertType to.
 * @returns {*} The converted type, or the given string if it couldn't be converted
 *   to the given type.
 * @ignore
 */
const convertType = exports.convertType = (value, type) => {
  if (!(0, _lang.isString)(value)) {
    throw new TypeError('Expected value to be a string.');
  }

  switch (resolveType(type)) {
    case BOOLEAN:
      return value.toLowerCase() === 'true';
    case INTEGER:
      return (0, _lang.toInteger)(value);
    case NUMBER:
      return (0, _lang.toNumber)(value);
    case STRING:
    default:
      return value;
  }
};
