// ghastly/lib/command/CommandGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _StringMap = require('../utils/StringMap');

var _StringMap2 = _interopRequireDefault(_StringMap);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @desc Groups commands.
 * @extends EventEmitter
 */
class CommandGroup extends _events2.default {
  /**
   * Constructor.
   * @param {string} name - The name of the group.
   */
  constructor(name) {
    super();

    /**
     * The name of the group.
     * @type {string}
     */
    this.name = name;

    /**
     * The commands that are a part of the group.
     * @type {StringMap.<CommandObject>}
     * @private
     */
    this.commands = new _StringMap2.default();

    /**
     * The middleware applied to the group.
     * @type {middlewareLayer[]}
     * @private
     */
    this.middleware = [];
  }

  /**
   * Adds a command to the group.
   * @param {CommandObject} command - The command.
   * @returns {CommandGroup} The instance this method was called on.
   */
  add(command) {
    const { name } = command;

    this.commands.set(name, command);
    command.linkGroup(this);

    return this;
  }

  /**
   * Emitted when middleware has been applied to commands in a group.
   * @event CommandGroup#middlewareUpdate
   * @param {middlewareLayer[]} layers - The layers which were applied.
   */

  /**
   * Adds the given middleware to the current group middleware layers.
   * @param {...middlewareLayer} layers - The layers to apply.
   * @returns {CommandGroup} The instance this method was called on.
   * @emits {CommandGroup#middlewareUpdate}
   */
  applyMiddleware(...layers) {
    this.middleware = [...layers, ...this.middleware];

    return this.emit('middlewareUpdate', this.middleware);
  }
}
exports.default = CommandGroup;
